From 30375877d981be8ede8620b13c6928342d929b58 Mon Sep 17 00:00:00 2001
From: Nick Wellnhofer <wellnhofer@aevum.de>
Date: Tue, 3 Jun 2025 15:50:54 +0200
Subject: [PATCH] parser: Fix custom SAX parsers without cdataBlock handler

Use characters handler if cdataBlock handler is NULL.

Regressed with 57e4bbd8. Should fix #934.
---
 parser.c | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/parser.c b/parser.c
index e6598413..c96c3bec 100644
--- a/parser.c
+++ b/parser.c
@@ -9502,12 +9502,11 @@ xmlParseCDSect(xmlParserCtxt *ctxt) {
      * OK the buffer is to be consumed as cdata.
      */
     if ((ctxt->sax != NULL) && (!ctxt->disableSAX)) {
-        if (ctxt->options & XML_PARSE_NOCDATA) {
-            if (ctxt->sax->characters != NULL)
-                ctxt->sax->characters(ctxt->userData, buf, len);
-        } else {
-            if (ctxt->sax->cdataBlock != NULL)
-                ctxt->sax->cdataBlock(ctxt->userData, buf, len);
+        if ((ctxt->sax->cdataBlock != NULL) &&
+            ((ctxt->options & XML_PARSE_NOCDATA) == 0)) {
+            ctxt->sax->cdataBlock(ctxt->userData, buf, len);
+        } else if (ctxt->sax->characters != NULL) {
+            ctxt->sax->characters(ctxt->userData, buf, len);
         }
     }
 
-- 
2.39.5

